### DISCUSSION TITLE
Is it possible to hide types from documentation?

### DISCUSSION THREAD

**Discussion by belak** [rust-lang/rust-bindgen#2513] [Q&A]
I'm working on an experimental legacy SDL crate and I noticed that bindgen creates both `SDL_Joystick` and `_STL_Joystick`, even though the latter shouldn't be a public-facing type at all. Is it possible to hide it from documentation without fully blacklisting it?

On a related note, is there a way to inject documentation from another source to these functions? It seems like most headers I've seen have no docs and lots of these functions don't have a great public documentation source any more.

**Comment by pvdrz** [comment]
> I'm working on an experimental legacy SDL crate and I noticed that bindgen creates both `SDL_Joystick` and `_STL_Joystick`, even though the latter shouldn't be a public-facing type at all. Is it possible to hide it from documentation without fully blacklisting it?

There's no way to inject such attributes right now. This seems like something that [`ParseCallbacks`](https://docs.rs/bindgen/latest/bindgen/callbacks/trait.ParseCallbacks.html) should do. One option would be to add a method to control the visibility of individual items. Another would be to add a method which allows to add arbitrary attributes. 

> On a related note, is there a way to inject documentation from another source to these functions? It seems like most headers I've seen have no docs and lots of these functions don't have a great public documentation source any more.

Yes, you can implement [`ParseCallbacks::process_comment`](https://docs.rs/bindgen/latest/bindgen/callbacks/trait.ParseCallbacks.html#method.process_comment). To inject additional stuff into the comments. This callback doesn't receive other info than the comment itself.

**Reply by belak** [reply]
I think ParseCallbacks::process_comment is close, but not quite what I'm looking for - I'm hoping to inject comments on methods and types where there aren't any, while this would let me modify existing comments.

**Reply by pvdrz** [reply]
Hmm yeah. Maybe introducing soemthing that lets you add arbitrary attributes could be used for that as `#[doc = "asd"]` is always an option.

What I have in mind is adding a method to `ParseCallbacks` like:

```rust
fn add_attributes(item_name: &str, item_kind: ItemKind) -> Vec<String> {
    if item_name == "_STL_Joystick" {
        vec!["doc(hidden)".to_owned()]
    } else if item_name == "ItemWithNoDocs" {
        vec!["doc =\"some fancy docs\""]
    } else {
        vec![]
    }
}
```
